### DISCUSSION TITLE
Example needed for .NET regarding authentication

### DISCUSSION THREAD

**Discussion by diegosasw** [scalar/scalar#3470] [Q&A] [answered]
I don't seem to be able to find samples with .NET

I am using Scalar UI with Microsoft OpenApi generated open api document. It works great for basic scenarios

Now I would like to explore possibilities for Bearer JWT secured endpoints.

I would like to be able to manually paste a JWT somewhere or, even better, have some automated authentication in which I can generate a mock JWT and pick it at Scalar UI. 

```
app.MapScalarApiReference(options =>
{
    options.Title = "Amosya API";
    options.Theme = ScalarTheme.None;
    // options.Authentication = new ScalarAuthenticationOptions // How to use?
 })
.AllowAnonymous();
```

The official website does not even mention .NET and I couldn't find anything in the forum search.

Any recommended sample code or repo where I could see advanced use cases for auth?

**Comment by xC0dex** [comment]
Which OpenAPI security scheme are you using, and which flows does it support? The authentication options are based on [this configuration](https://github.com/scalar/scalar/blob/main/documentation/configuration.md#authentication-partial).

**Reply by diegosasw** [reply]
I´m using Bearer token (OAuth2/OpenId Connect provided by Cognito hosted UI in a different App, and being received as Bearer token in this .NET API where I have Scalar).
> The authentication options are based on [this configuration](https://github.com/scalar/scalar/blob/main/documentation/configuration.md#authentication-partial).

Is there any .NET C# sample on how to use it? Those schemas are confusing. For instance:
- The first one mentions ApiKey but it's not what I'm looking for (Bearer token is not the same as Basic token or any other api key in other http header)
- The second one mentions oauth2 and requires `clientId` but no secret?
- And how would I proceed in C# if I simply want to select a prefilled mocked JWT to be used as http header `Authorization: Bearer <jwt_here>` ? That way, instead of attempting any kind of authentication interactive flow (which I don't want), I can just preselect a mocked JWT that my App could produce at runtime, or similar.

**Reply by xC0dex** [reply, marked as answer]
Hey @diegosasw,

Thank you for your patience, I was quite busy this week. The Scalar team is currently improving the whole OAuth experience. Typically, Scalar checks the OpenApi document for the security schemes. The scheme defines if the API requires an API key or a bearer token. It also contains information about the authentication flow that should be used. 
Currently, it's only possible to prefill the clientId in the UI. you have to provide the secret manually.
TBH, I'm not sure how to set up the OpenAPI document to use a mocked JWT instead of using one of the flows. I would say you _could_ try out the API Key approach and manually set the value in the options to `bearer <token>`.

**Reply by diegosasw** [reply]
Thanks.
I can see that, as soon as my OpenApi document has the security schemes, Scalar displays options for Bearer token authentication properly.

Just in case this is useful to others, I am using .NET 9 and nuget `Microsoft.AspNetCore.OpenApi` to automatically generate an OpenApi document, and then I use a document transformer to add the security scheme. See my DI container registration method

```
public static class OpenApiRegistrations
{
    internal static IServiceCollection AddMicrosoftOpenApi(
        this IServiceCollection services)
    {
        services.AddOpenApi(options =>
        {
            options.OpenApiVersion = OpenApiSpecVersion.OpenApi3_0;
            options.AddDocumentTransformer<OpenApiSecuritySchemeTransformer>();
        });
        
        return services;
    }
}

public class OpenApiSecuritySchemeTransformer
    : IOpenApiDocumentTransformer
{
    public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context,
        CancellationToken cancellationToken)
    {
        document.Info.Title = Constants.OpenApi.Title;
        document.Info.Description = Constants.OpenApi.Description;
        document.Info.Contact = new OpenApiContact
        {
            Name = Constants.OpenApi.Name,
            Email = Constants.OpenApi.Email,
            Url = new Uri(Constants.OpenApi.Website)
        };

        var securitySchema =
            new OpenApiSecurityScheme
            {
                Type = SecuritySchemeType.Http,
                Scheme = "bearer",
                BearerFormat = "JWT",
                Description = "JWT Authorization header using the Bearer scheme."
            };

        var securityRequirement =
            new OpenApiSecurityRequirement
            {
                {
                    new OpenApiSecurityScheme
                    {
                        Reference = new OpenApiReference
                        {
                            Id = Constants.OpenApi.SecuritySchemeBearer,
                            Type = ReferenceType.SecurityScheme
                        }
                    },
                    []
                }
            };
        
        document.SecurityRequirements.Add(securityRequirement);
        document.Components = new OpenApiComponents()
        {
            SecuritySchemes = new Dictionary<string, OpenApiSecurityScheme>()
            {
                { Constants.OpenApi.SecuritySchemeBearer, securitySchema }
            }
        };
        return Task.CompletedTask;
    }
}
```

Then, in my HTTP pipeline I simply add Scalar UI pointing to my OpenApi document path. See my extension method

```
public static WebApplication UseOpenUiWithScalar(
    this WebApplication app)
{
   app
            .MapOpenApi()
            .AllowAnonymous();

    app
        .MapScalarApiReference(options =>
        {
            options.WithOpenApiRoutePattern("/openapi/v1.json");
            options.Theme = ScalarTheme.None;
            options.Authentication = 
                new ScalarAuthenticationOptions
                {
                    PreferredSecurityScheme = Constants.OpenApi.SecuritySchemeBearer
                };
        })
        .AllowAnonymous();

    return app;
}
```

More info about OpenApi generation for .NET 9 [here](https://github.com/dotnet/aspnetcore/issues/39761#issuecomment-2439607317)

**Reply by xC0dex** [reply]
Hey @diegosasw,

Glad that you found a solution. Just to let you know that I'm currently working on a library called [`APIWeaver`](https://github.com/xC0dex/APIWeaver) that provides some extension methods for the built-in generator, for instance for authentication:

```c#
builder.Services.AddOpenApi(options =>
{
    // Adds a security scheme to the OpenAPI document
    // and update all operations with the security requirement
    options.AddSecurityScheme("Bearer", scheme =>
    {
        scheme.In = ParameterLocation.Header;
        scheme.Type = SecuritySchemeType.OAuth2;
        scheme.Flows = new OpenApiOAuthFlows
        {
            ClientCredentials = new OpenApiOAuthFlow
            {
                TokenUrl = new Uri("https://localhost:5001/oauth2/token")
            }
        };
    });

    // Adds 401 and 403 responses to the operations
    options.AddAuthResponse();
});
```

**Reply by diegosasw** [reply]
@xC0dex it looks very useful. Will give it a try! Thanks

**Reply by bislerium** [reply]
> options.Authentication = 
>                 new ScalarAuthenticationOptions
>                 {
>                     PreferredSecurityScheme = Constants.OpenApi.SecuritySchemeBearer
>                 };

is this really needed, since the openapi spec already contains the jwt bearer security requirement?
